src/reducers/FormReducer.js

```diff
@@ -0,0 +1,58 @@
+/**
+ * mSupply Mobile
+ * Sustainable Solutions (NZ) Ltd. 2019
+ */
+
+import { FORM_ACTIONS } from '../actions/FormActions';
+
+const initialState = config => {
+  if (!config) return {};
+
+  return config.reduce(
+    (acc, { key, initialValue, validator, isRequired }) => ({
+      ...acc,
+      [key]: {
+        value: initialValue,
+        isValid: validator ? validator(initialValue) : true,
+        hasValue: isRequired ? !!initialValue : true,
+        validator,
+        isRequired,
+      },
+    }),
+    {}
+  );
+};
+
+export const FormReducer = (state = initialState(), action) => {
+  const { type } = action;
+
+  switch (type) {
+    case FORM_ACTIONS.INITIALISE: {
+      const { payload } = action;
+      const { config } = payload;
+      return initialState(config);
+    }
+    case FORM_ACTIONS.UPDATE: {
+      const { payload } = action;
+      const { field, value } = payload;
+
+      const stateData = state[field];
+
+      const { isRequired, validator } = stateData;
+
+      const newStateData = {
+        ...stateData,
+        value,
+        hasValue: isRequired ? !!value : true,
+        isValid: validator ? validator(value) : true,
```

Comment:
Re a comment I made elsewhere: I'm wondering what value `hasValue` adds - and whether it would be better to have:
```
      const isValid = (isRequired ? !!value : true) && (validator ? validator(value) : true);

      const newStateData = {
        ...stateData,
        value,
        isValid,
      };
```

Reply:
db050fbe668dd962dca695f1c0b72f12cf35aff0

You got me thinking.. hopefully that's a good thing..

I've just used a validator function for the required fields. You're right, a field being required really is just another variable in the 'validity' of that field, so it does seem right to just use the same 'validator' concept for all of this, leaving the `isRequired` field really just for a UI hint. 

Hopefully this is better?